android/src/main/java/com/virgilsecurity/rn/crypto/RNVirgilCryptoModule.java

```diff
@@ -310,6 +319,77 @@ public WritableMap decryptAndVerifyDetached(String dataBase64,
     }
   }
 
+  @ReactMethod(isBlockingSynchronousMethod = true)
+  public WritableMap generateGroupSession(String groupIdBase64) {
+      byte[] sessionId = this.crypto.computeHash(
+              Encodings.decodeBase64(groupIdBase64),
+              HashAlgorithm.SHA512
+      );
+      sessionId = Arrays.copyOfRange(sessionId, 0, 32);
+      GroupSessionTicket initialEpochTicket = new GroupSessionTicket();
+      initialEpochTicket.setRng(this.crypto.getRng());
+      try {
+        initialEpochTicket.setupTicketAsNew(sessionId);
+      } catch (FoundationException e) {
+          return ResponseFactory.createErrorResponse(e);
+      }
+
+      GroupSessionMessage initialEpochMessage = initialEpochTicket.getTicketMessage();
+      int epochNumber = (int)initialEpochMessage.getEpoch();
+      byte[] data = initialEpochMessage.serialize();
+
+      WritableArray epochMessages = Arguments.createArray();
+      epochMessages.pushString(Encodings.encodeBase64(data));
+
+      WritableMap responseMap = Arguments.createMap();
+      responseMap.putString("sessionId", Encodings.encodeBase64(sessionId));
+      responseMap.putInt("currentEpochNumber", epochNumber);
+      responseMap.putArray("epochMessages", epochMessages);
+      return ResponseFactory.createMapResponse(responseMap);
+  }
+
+  @ReactMethod(isBlockingSynchronousMethod = true)
+  public WritableMap importGroupSession(ReadableArray epochMessagesBase64) {
+      List<GroupSessionMessage> epochMessages = new ArrayList<>(epochMessagesBase64.size());
+      try {
+          for(Object epochMessageBase64 : epochMessagesBase64.toArrayList()) {
+              epochMessages.add(GroupSessionMessage.deserialize(Encodings.decodeBase64((String)epochMessageBase64)));
+          }
+      } catch (FoundationException e) {
+          return ResponseFactory.createErrorResponse(e);
+      }
+
+      Collections.sort(epochMessages, new Comparator<GroupSessionMessage>() {
+          @Override
+          public int compare(GroupSessionMessage a, GroupSessionMessage b) {
+              return (int)(a.getEpoch() - b.getEpoch());
+          }
+      });
+
+      WritableArray serializedEpochMessages = Arguments.createArray();
+
+      GroupSession session = new GroupSession();
+      session.setRng(this.crypto.getRng());
+
+      for(GroupSessionMessage epochMessage : epochMessages) {
```

Comment:
Is there any limit for number of epochs here?

Reply:
Not here. Limiting the number of epochs will be part of E3kit implementation.